### PR TITLE
feat: Implement Jaeger Propagator.

### PR DESCRIPTION
**Pull request by fhwang**
I think this is ready for review now. I started by mimicking the B3 propagator and changing where needed. A few design notes:

- The Jaeger format supports baggage using keys like `uberctx-{baggage key}` which then begs the question as to how to customize the Rack extractor and injector. I chose to go with a design that allowed optional blocks for handling keys.
- Regarding testing this baggage functionality, I opted to include SDK as a development dependency. I spent some time trying to do some surface-level mocking with Mocha, but it ended up being pretty tangled-up. I think the ideal way to test this is to have something that is a functional baggage implementation, which seems like a good use of the SDK. If we consider this a risk I can make another pass on using mocking however.

### CODE REVIEW

**Review by mwear — COMMENTED** [review]
@fhwang sorry for the turnaround time on this review. I took a look over this last week and realized that some of the issues you were running into processing baggage headers with Rack could likely be improved with some changes to our Getter implementation.

I started experimenting with that and opened #589. With the Getter changes, we are able to write the injectors and extractors in a carrier agnostic way. We no longer need to specify Rack specific keys, or maintain Rack specific injector / extractor instances. If we need to extract from a Rack based carrier we just pass in a RackEnvGetter.

This generally looks to be on the right track. If you get a chance, can you rebase this, take a look at the updated B3 injectors / extractor, and follow the same pattern?

**Comment by fhwang** [comment]
No problem. I'll take a look at incorporating #589 in the next day or so.

**Comment by fhwang** [comment]
Closes #470.

File: propagator/jaeger/README.md
Comment (fbogsany) [inline propagator/jaeger/README.md:5]:
```suggestion
conjunction with each other as well as with injectors and extractors for
```

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:20]: The Jaeger spec (https://www.jaegertracing.io/docs/1.21/client-libraries/#propagation-format) seems to allow trace ids to be lengths other than 16 or 32 hex chars.

trace id:

> Receivers MUST accept hex-strings shorter than 32 characters and 0-pad them on the left

span id:

> Receivers MUST accept hex-strings shorter than 16 characters and 0-pad them on the left

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:21]: You shouldn't need to enumerate these values. Just parse the flags, convert to an integer, and use a bitwise & to test for sampled or debug.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:76]: Usually, if you can get away with `String#start_with?` it's more efficient than a regex. I think you can use that here. Is there conversion of '_' to '-' necessary? I believe the RackGetter handles this already.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb:13]: Any reason not to represent these as base16 like the spec: https://www.jaegertracing.io/docs/1.21/client-libraries/#propagation-format

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:79]: I believe that baggage values should be url decoded on extract: https://www.jaegertracing.io/docs/1.21/client-libraries/#value-encoding

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb:50]: Similarly, I think that baggage values should be url encoded on inject (https://www.jaegertracing.io/docs/1.21/client-libraries/#value-encoding)

File: api/lib/opentelemetry/context/propagation/rack_env_setter.rb
Comment (mwear) [inline api/lib/opentelemetry/context/propagation/rack_env_setter.rb:13]: It seems like this is only used it tests. I fear it might cause confusion for users and I don't know that there is much of a real world use case for it. My preference would be to omit it and create the rack formatted keys by hand in tests.

**Review by mwear — CHANGES_REQUESTED** [review]

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (fhwang) [reply propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:79]: That part of the spec says to only URL-encode when using the `HTTP_HEADERS` propagation format wrt OpenTracing. I'm not completely clear how we'd map this part of the spec to OpenTelemetry, but maybe we should just use the `TEXT_MAP` format instead?

**Comment by fhwang** [comment]
Thanks for the helpful review @mwear . I believe I've addressed all the issues, with the exception of URL encoding / decoding. I'm submitting these changes as standalone commits to ease review but once I have approval I was thinking of squash-merging to keep `main` clean.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:20]: This regex enforces that the trace id and span id have an even number of hex digits. I don't think that is mandated by the Jaeger spec and we should probably allow for an odd number of digits.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/operation.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/operation.rb:20]: There's really no reason for this class. You can put these constants on the Jaeger module or in the injector / extractor depending on whether they need to be shared in both places.

**Review by mwear — CHANGES_REQUESTED** [review]

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger.rb:31]: This and the constants below can be private

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:54]: This regex can be a constant

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:84]: CGI.escape and CGI.unescape are what we should be using here
```suggestion
            value = CGI.escape(raw_value)
```

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:102]: I think this can be simplified to:
```suggestion
          span_id_str.rjust(16, '0')
```

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:107]: Similarly, I think this can be:
```suggestion
         trace_id.rjust(32, '0')
```

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb:49]:
```suggestion
            encoded_value = CGI.escape(value)
```

**Review by mwear — CHANGES_REQUESTED** [review]

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (fhwang) [reply propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:107]: The current implementation emits 64-bit trace IDs when it receives 64-bit trace IDs; using `rjust(32, '0')` would force them to zero-padded 128-bit trace IDs. Both would seem to match the Jaeger spec, but given that the spec says "Clients in some languages support 128-bit, migration pending", that would imply to me that some clients **do not** support 128-bit. Meaning that it might be marginally safer to stick with 64-bit trace IDs in case there are legacy clients downstream.

Or is that gold-plating? Happy to just change it too.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger.rb
Comment (fhwang) [reply propagator/jaeger/lib/opentelemetry/propagator/jaeger.rb:31]: These are private now.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb
Comment (mwear) [reply propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb:49]: The official go jaeger client library uses [url.QueryEscape](https://golang.org/pkg/net/url/#QueryEscape) which works the same as CGI.escape, see: https://github.com/jaegertracing/jaeger-client-go/blob/master/propagation.go#L76-L92. The Ruby Jaeger client also uses `CGI.escape`, see: https://github.com/salemove/jaeger-client-ruby/blob/master/lib/jaeger/injectors.rb#L28.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [reply propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:107]: The [opentelemetry spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spancontext) says:

> TraceId A valid trace identifier is a 16-byte array with at least one non-zero byte.

So, I think when extracting we can safely zero pad to 32 hex chars.

Comment (fhwang) [reply propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:107]: Sounds good, I'll zero-pad to 32 hex chars.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb
Comment (fhwang) [reply propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb]: 👍 works for me, I've changed these to CGI escape and unescape.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb:37]:
```suggestion
        def inject(carrier, context, setter = nil)
```

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb:34]: Docs are out of date here too. The setter is no longer a block. Take a look at the docs for other injectors.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:47]: Docs are out of date. The getter is no longer a block. Take a look at the docs for other extractors.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_injector.rb:58]: This constant isn't found and needs further qualification. Its full path is `OpenTelemetry::Trace::TraceFlags`, but this will do:
```suggestion
          if span_context.trace_flags == Trace::TraceFlags::SAMPLED
```

**Review by mwear — CHANGES_REQUESTED** [review]
This is almost here. I did an end to end test earlier and found a couple of issues. Two are related directly to this PR and one is an issue with the RackEnvGetter. I'll open another PR to fix it. Also, the docs are slightly out of date and also noted in the comments.

File: propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb
Comment (mwear) [inline propagator/jaeger/lib/opentelemetry/propagator/jaeger/text_map_extractor.rb:86]: This will create a new baggage object for each value set. You can more efficiently make multiple modifications to baggage by using the `build` method. Basically, I think you can wrap this whole method in a:

```ruby
OpenTelemetry.baggage.build(context) do |b|
 ...
  b.set_value(baggage_key, value)
  ...
end
```

**Comment by fhwang** [comment]
@mwear Ready for another review. Thanks!

**Review by mwear — APPROVED** [review]
Looks good! Thanks @fhwang!

**Pull request merged**
